File: app/src/main/java/com/example/pokedex/data/local/entity/DetailPokemonEntity.java
package com.example.pokedex.data.local.entity;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.PrimaryKey;


@Entity(tableName = "detail_pokemon_table")
public class DetailPokemonEntity {
    @ColumnInfo(name = "caught")
    private boolean caught = false;

    @PrimaryKey
    @NonNull
    private String id;

    private String name;

    private String nickname;

    private int height;

    private int weight;

    private String url;

    @NonNull
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getHeight() {
        return height;
    }

    public void setHeight(int height) {
        this.height = height;
    }

    public int getWeight() {
        return weight;
    }

    public void setWeight(int weight) {
        this.weight = weight;
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public boolean isCaught() {
        return caught;
    }

    public void setCaught(boolean caught) {
        this.caught = caught;
    }

    public String getNickname() {
        return nickname;
    }

    public void setNickname(String nickname) {
        this.nickname = nickname;
    }

    public DetailPokemonEntity(boolean caught, @NonNull String id, String name, int height, int weight, String url) {
        this.caught = caught;
        this.id = id;
        this.name = name;
        this.height = height;
        this.weight = weight;
        this.url = url;
    }
}

File: core/src/main/java/com/example/data/local/entity/DetailPokemonEntity.kt
package com.example.data.local.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.example.data.remote.response.PokemonStatDto
import com.example.data.remote.response.PokemonStatsDto
import com.example.domain.model.DetailPokemonModel
import com.example.domain.model.PokemonStatModel
import com.example.domain.model.PokemonStatsModel

@Entity(tableName = "detail_pokemon_table")
data class DetailPokemonEntity(
  @ColumnInfo(name = "caught")
  var isCaught: Boolean = false,
  @PrimaryKey
  var id: String,
  var name: String? = null,
  var height: Int = 0,
  var weight: Int = 0,
  var url: String? = null,
  var nickname: String? = null,
  val stats: List<PokemonStatsEntity>
) {
  companion object {
    fun transform(entity: DetailPokemonEntity): DetailPokemonModel {
      val transformedList = arrayListOf<PokemonStatsModel>()
      entity.stats.forEach {
        transformedList.add(PokemonStatsEntity.transform(it))
      }
      return DetailPokemonModel(
        false,
        entity.id,
        entity.name,
        entity.height,
        entity.weight,
        entity.url,
        entity.nickname,
        transformedList
      )
    }
  }

}

data class PokemonStatsEntity(
  val baseStats: Int? = null,
  val stat: PokemonStatEntity? = null
) {
  companion object {
    fun transform(entity: PokemonStatsEntity): PokemonStatsModel {
      return PokemonStatsModel(
        entity.baseStats,
        entity.stat?.let { PokemonStatEntity.transform(it) }
      )
    }
  }
}

data class PokemonStatEntity(
  val name: String? = null,
  val url: String? = null
) {
  companion object {
    fun transform(entity: PokemonStatEntity): PokemonStatModel {
      return PokemonStatModel(
        entity.name,
        entity.url
      )
    }
  }
}
